File: Easy/0020_valid-parentheses/valid-parentheses.py
#!/bin/python3

class Solution:
    def isValid(self, s: str) -> bool:
        pair : dict[str, str] = {
            ')': '(',
            '}': '{',
            ']': '[',
        }
        stack = []
        for c in s:
            if c in "({[":
                stack.append(c)
            elif not stack or stack.pop() != pair[c]:
                    return False
        return not stack


def main():
    tests : list[str] = {
        "()",
        "()[]{}",
        "(]",
        "([])",
        "([)]"
    }

    for t in tests:
        print(t,"=> ", Solution().isValid(t))

if __name__ == "__main__":
    main()

File: Easy/0020_valid-parentheses/valid-parentheses.go
package main

import (
	"fmt"
)

func isValid(s string) bool {
  pair := map[rune]rune{
		')': '(',
		'}': '{',
		']': '[',
	}

	stack := make([]rune, 0, len(s))

	for _, ch := range s {
		switch ch {
			case '(', '[', '{':
				stack = append(stack, ch)

			case ')', ']', '}':
				if len(stack) == 0 || stack[len(stack) - 1] != pair[ch] {
					return false
				}
				stack = stack[:len(stack) - 1]

			default:
				return false
		}
	}
	return len(stack) == 0
}

func main() {
	tests := []string{
		"()",
		"()[]{}",
		"(]",
		"([])",
		"([)]",
		"",
	}

	for _, t := range tests {
		fmt.Printf("%q => %v\n", t, isValid(t))
	}
}
